### PR TITLE
Add the Synapse function `types.map_username_to_mxid_localpart` to the Module API

### PR DESCRIPTION
**Pull request by H-Shay**
As the title states. It will be helpful to be able to use this in the LDAP module and others might find it useful as well.

### CODE REVIEW

File: tests/module_api/test_api.py [hunk 1]
```diff
@@ -635,6 +635,14 @@ def test_check_push_rules_actions(self) -> None:
...
         )
 
+    def test_normalize_username(self) -> None:
+        username = "Haxxor"
+        username2 = "_leet"
+        username3 = "aNoThErTeSt"
+        self.assertEqual(self.module_api.normalize_username(username), "haxxor")
+        self.assertEqual(self.module_api.normalize_username(username2), "=5fleet")
+        self.assertEqual(self.module_api.normalize_username(username3), "anothertest")
+
```
Comment (DMRobertson) [inline tests/module_api/test_api.py:645]: Given that we're re-exposing one function withouot any I'm not sure how much this test gives us. I suppose it does force us to check that we continue to expose this in the module API!

File: synapse/module_api/__init__.py [hunk 2]
```diff
@@ -573,6 +574,23 @@ def get_qualified_user_id(self, username: str) -> str:
             return username
         return UserID(username, self._hs.hostname).to_string()
 
+    def normalize_username(
+        self, username: Union[str, bytes], case_sensitive: bool = False
+    ) -> str:
+        """Map a username onto a string suitable for a MXID
```
Comment (DMRobertson) [inline synapse/module_api/__init__.py:580]: Maybe add a hyperlink to the spec here? It's not just that we've got a reasonable-looking mapping, but that it's the specific one defined in the spec.

**Review by DMRobertson — COMMENTED** [review]
I'm not convinced (-0.3) that Synapse's module API is the right place for this, rather than `matrix-common`. This isn't a tool that helps modules interact with Synapse; it's a tool that helps modules run a standard algorithm defined by the spec. (I agree that `matrix-common` says it's for Synapse, Sygnal and Sydent, but I think its purpose is "general-purpose utilities for working with the matrix spec".)

That said, it might be that we've done this kind of thing in the module API before. I'm happy to defer to the opinion of @babolivier as chief module-wrangler. How does that sound, Shay? Either way, I've left some thoughts on the changes too.

File: tests/module_api/test_api.py (hunk 1 above)
Comment (H-Shay) [reply tests/module_api/test_api.py:645]: You know I wasn't so sure of the value of these either but as I junior I try to adhere to a "test everything if possible" philosophy. Happy to dump them if that's cleaner.

File: synapse/module_api/__init__.py (hunk 2 above)
Comment (H-Shay) [reply synapse/module_api/__init__.py:580]: Good idea, sorted.

**Comment by H-Shay** [comment]
> How does that sound, Shay?

Fine with me, I don't have much of an opinion on where this should go-happy to defer to @babolivier on this one.

**Comment by babolivier** [comment]
I think it belongs neither in the module API nor in `matrix-common`. `matrix-common` is, as the name says, specific to Synapse, Sydent and Sygnal (and we've been pretty strict about this in the past), and we should not put stuff in there that e.g. third-party module authors would see as a signal that they should use `matrix-common` in their projects.

Ideally we should have a place for this kind of things to go, but we don't currently have a library with sufficient guarantees about stability and backwards-compatibility where this could go. Historically, for small functions like this that are trivial to reproduce, we've just been copying or rewriting them into the module directly, so that's probably what we should do here.

**Pull request closed without merging**